core/src/main/scala/org/apache/spark/storage/PushBasedFetchHelper.scala

```diff
@@ -0,0 +1,289 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.storage
+
+import java.util.concurrent.TimeUnit
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.util.{Failure, Success}
+
+import org.roaringbitmap.RoaringBitmap
+
+import org.apache.spark.MapOutputTracker
+import org.apache.spark.MapOutputTracker.SHUFFLE_PUSH_MAP_ID
+import org.apache.spark.internal.Logging
+import org.apache.spark.network.buffer.ManagedBuffer
+import org.apache.spark.network.shuffle.{BlockStoreClient, MergedBlockMeta, MergedBlocksMetaListener}
+import org.apache.spark.storage.BlockManagerId.SHUFFLE_MERGER_IDENTIFIER
+import org.apache.spark.storage.ShuffleBlockFetcherIterator._
+
+/**
+ * Helper class for [[ShuffleBlockFetcherIterator]] that encapsulates all the push-based
+ * functionality to fetch merged block meta and merged shuffle block chunks.
+ */
+private class PushBasedFetchHelper(
+   private val iterator: ShuffleBlockFetcherIterator,
+   private val shuffleClient: BlockStoreClient,
+   private val blockManager: BlockManager,
+   private val mapOutputTracker: MapOutputTracker) extends Logging {
+
+  private[this] val startTimeNs = System.nanoTime()
+
+  private[this] val localShuffleMergerBlockMgrId = BlockManagerId(
+    SHUFFLE_MERGER_IDENTIFIER, blockManager.blockManagerId.host,
+    blockManager.blockManagerId.port, blockManager.blockManagerId.topologyInfo)
+
+  /** A map for storing merged block shuffle chunk bitmap */
+  private[this] val chunksMetaMap = new mutable.HashMap[ShuffleBlockChunkId, RoaringBitmap]()
```

Comment:
`fetchMergedLocalBlock` is not invoked from the task thread right (from `fetchMergedLocalBlocks`) ?
That updates `chunksMetaMap` ?

Reply:
@mridulm That's a very good catch. Sorry, missed that. Netty thread will be adding to it for the local block. I will work on fixing this.

Reply:
I have changed this to a `ConcurrentHashMap` and switched the order of these lines in `fetchMergedLocalBlock`
```
        chunksMetaMap.put(shuffleChunkId, chunksMeta(chunkId))
        iterator.addToResultsQueue(
          SuccessFetchResult(shuffleChunkId, SHUFFLE_PUSH_MAP_ID, blockManagerId, buf.size(), buf,
            isNetworkReqDone = false))
```
I think this should solve any issues because all we want is the guarantee that for a `SuccessFetchResult` related to a shuffleChunkId, the meta for that shuffleChunkId is present in `chunksMetaMap`. 
I will think a bit more about this tomorrow.

Reply:
I took another look at it and I think making it a `ConcurrentMap` and swapping the order of those lines fix any issues related to multiple thread. PTAL.

Reply:
@mridulm In the new changes, I have moved the part in `fetchMergedLocalBlock` which was modifying the map to be handled just by the task thread when it processes `PushMergedLocalMetaFetchResult`. So, I have changed this map back to a regular hashmap because now it just gets modified by task thread. PTAL